### ISSUE TITLE
Using pdfium with axum

### ISSUE THREAD

**Issue by danwritecode** [ajrcarey/pdfium-render#172]
Hey there! This crate has been amazing and it's allowed me to do exactly what i need, however, I'm using pdfium in a web server environment with Axum and I run into the below issue. 

```
rustc: `*mut pdfium_render::bindgen::fpdf_document_t__` cannot be sent between threads safely
 within `impl std::future::Future<Output = Result<(), AppError>>`, the trait `Send` is not implemented for `*mut pdfium_render::bindgen::fpdf_document_t__`, which is required by `impl std::future::Future<Output = Result<(), AppError>>: Send` [E0277]
```

I have "sync" enabled:
`pdfium-render = { version = "0.8.26", features = ["sync"] }`

Currently exploring a different solution entirely, but pdfium is by far the most robust solution available in the rust ecosystem and this library is really a joy to use. 

Appreciate any guidance!

**Comment by ajrcarey** [comment]
Hi @danwritecode , thank you for reporting the issue. I know you already know this, but I feel compelled to reiterate that Pdfium is not thread-safe. So my instinct is that it's probably not possible to _safely_ do what you're trying to do.

That said, if you want to try and do it anyway, I suggest the following `git clone`-ing this repo, adding `Sync` and `Send` impls for `FPDFDOCUMENT` to `src/pdf/document.rs` as per below, and trying it out in your environment. I'd be curious as to whether it works or not. If it does, then we could at merging those additional impls into this repo.

Currently, the very bottom of `document.rs` has:

```
#[cfg(feature = "sync")]
unsafe impl<'a> Sync for PdfDocument<'a> {}

#[cfg(feature = "sync")]
unsafe impl<'a> Send for PdfDocument<'a> {}
```

You want to try extending this to also include:

```
#[cfg(feature = "sync")]
unsafe impl<'a> Sync for FPDFDOCUMENT {}

#[cfg(feature = "sync")]
unsafe impl<'a> Send for FPDFDOCUMENT {}
```

(FPDFDOCUMENT is just a type alias for fpdf_document_t)

I will leave this issue open for the next week or so to give you a chance to try it out and report back, if you wish.

**Comment by ajrcarey** [comment]
(PS it's possible, perhaps likely, that this will result in a cascade where Sync and Send also need to be implemented for child types of `FPDFDOCUMENT` such as `FPDFPAGE`, etc.)

**Comment by danwritecode** [comment]
Awesome, I appreciate the response! I will fork it and give it a shot and report back with my findings.

**Comment by ajrcarey** [comment]
Even if it doesn't work (and if I had to guess, I'd guess it won't), you may be able to work around it by lifting your `PdfDocument` into a `OnceCell` or `Mutex` or similar and accessing it that way using a more traditional locking approach rather than passing it around as a function argument.

**Comment by danwritecode** [comment]
Finally got a chance to look at this. I run into this issue immediately: 

```
error[E0751]: found both positive and negative implementation of trait `Send` for type `*mut fpdf_document_t__`:
   --> src/pdf/document.rs:432:1
    |
432 | unsafe impl<'a> Send for FPDF_DOCUMENT {}
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ positive implementation here
    |
    = note: negative implementation in crate `core`

error[E0117]: only traits defined in the current crate can be implemented for arbitrary types
   --> src/pdf/document.rs:432:1
    |
432 | unsafe impl<'a> Send for FPDF_DOCUMENT {}
    | ^^^^^^^^^^^^^^^^^^^^^^^^^-------------
    | |                        |
    | |                        `*mut fpdf_document_t__` is not defined in the current crate because raw pointers are always foreign
    | impl doesn't use only types from inside the current crate
    |
    = note: define and implement a trait or new type instead
help: consider introducing a new wrapper type
    |
432 + struct WrapperType(*mut fpdf_document_t__);
433 + 
434 ~ unsafe impl<'a> Send for WrapperType {}
```

I attempted to wrap it in a wrapper type, and using the create locally using path in my cargo.toml, but still run into the same error as referenced originally. 

Going to look at using something like once_cell and wrapping pdfium in a mutex. Probably not ideal but maybe fine for now, we'll see. I'll report back on that approach as well.

**Comment by danwritecode** [comment]
I got things working! Here is what i used: 


```rust
static PDFIUM: OnceCell<Mutex<Pdfium>> = OnceCell::const_new();

async fn init_pdfium() -> Mutex<Pdfium> {
    let pdfium = Pdfium::new(
        match Pdfium::bind_to_system_library() {
            Ok(p) => p,
            Err(e) => {
                panic!("Failed to init Pdfium | Error: {}", e)
            }
        }
    );

    Mutex::new(pdfium)
}
```

Then in my client lib:
```rust
pub struct PdfClient<'a> {
    pdfium: &'a Mutex<Pdfium>,
    bytes: Vec<u8>
}
```

From there i was able to call `lock().await` just fine.

**Comment by ajrcarey** [comment]
That's excellent news, @danwritecode ! Great work.

If you had a small code sample that demonstrated this, I'd be happy to write it up and add it to the` /examples` in this repo. I'm sure there are others who could benefit from your work in the future...?

**Comment by danwritecode** [comment]
@ajrcarey sounds good! I will create a working example some time this week. Do you want a full axum app? I will keep it as minimal as possible.

**Comment by ajrcarey** [comment]
I think it depends on how much effort a "full" axum app would take. (I'm aware of axum but haven't used it, so unfortunately I can't offer much guidance.) If it's easy to make a minimal axum app and show the use of `pdfium-render `with `OnceCell`, that would be ideal. But if it's a lot of work, let's find another way. Ideally it would be something you only needed to spend a few minutes on, rather than hours.

**Comment by danwritecode** [comment]
It will only take me a few minutes, no worries at all. I'll PR it into the examples folder this week.

**Comment by ajrcarey** [comment]
Superb, thank you so much @danwritecode . The example looks great. I may make some small adjustments to the comments - hope that's ok. Could you go into a little more detail about this one?

> SCOPING PDFIUM: You may want to scope access to PDFIUM that way the mutex guard is dropped asap

I understand scoping, but is this something specific you had to for Axum?

**Comment by danwritecode** [comment]
> Superb, thank you so much @danwritecode . The example looks great. I may make some small adjustments to the comments - hope that's ok. Could you go into a little more detail about this one?
> 
> 
> 
> > SCOPING PDFIUM: You may want to scope access to PDFIUM that way the mutex guard is dropped asap
> 
> 
> 
> I understand scoping, but is this something specific you had to for Axum?

Feel free to either make the tweaks or suggest them!

Regarding that comment it wasn't something i had to do explicitly for it to work with axum, but I felt it was helpful for the user to be aware of. Though if you feel it's unrelated then we can remove it.

**Comment by ajrcarey** [comment]
Merged PR. Adjusted example to use an HTTP GET rather than POST for easier direct testing. Updated comments. Updated dev-dependencies in `Cargo.toml`. Updated `README`.

The example will be included in the release of crate version 0.8.27.

**Comment by danwritecode** [comment]
awesome thank you! was traveling, so appreciate your help with the finishing touches.
